orchagent/mplsrouteorch.cpp

```diff
@@ -142,74 +146,97 @@ void RouteOrch::doLabelTask(Consumer& consumer)
                     if (fvField(i) == "mpls_pop")
                         pop_count = to_uint<uint8_t>(fvValue(i));
 
-                    if (fvField(i) == "blackhole")
-                        blackhole = fvValue(i) == "true";
+                    if (fvField(i) == "nexthop_group")
+                        nhg_index = fvValue(i);
                 }
-                vector<string> ipv = tokenize(ips, ',');
-                vector<string> alsv = tokenize(aliases, ',');
-                vector<string> mpls_nhv = tokenize(mpls_nhs, ',');
 
-                /* Resize the ip vector to match ifname vector
-                 * as tokenize(",", ',') will miss the last empty segment. */
-                if (alsv.size() == 0 && !blackhole)
+                SWSS_LOG_INFO("Label route %u has nexthop_group: %s, ips: %s, "
+                              "MPLS NHs: %s, aliases: %s",
+                                    label,
+                                    nhg_index.c_str(),
+                                    ips.c_str(),
+                                    mpls_nhs.c_str(),
+                                    aliases.c_str());
+
+                /*
+                 * A route should not fill both nexthop_group and ips /
+                 * aliases.
+                 */
+                if (!nhg_index.empty() && (!ips.empty() || !aliases.empty()))
                 {
-                    SWSS_LOG_WARN("Skip the route %s, for it has an empty ifname field.", key.c_str());
+                    SWSS_LOG_ERROR("Route %s has both nexthop_group and ips/aliases",
+                                    key.c_str());
                     it = consumer.m_toSync.erase(it);
                     continue;
                 }
-                else if (alsv.size() != ipv.size())
+
+                ctx.nhg_index = nhg_index;
+
+                vector<string> ipv;
+                vector<string> alsv;
+                vector<string> mpls_nhv;
+
+                /*
+                 * If the nexthop_group is empty, create the next hop group key
+                 * based on the IPs and aliases.  Otherwise, get the key from
+                 * the NhgOrch.
+                 */
+                if (nhg_index.empty() && aliases.length() == 0)
                 {
-                    SWSS_LOG_NOTICE("Route %s: resize ipv to match alsv, %zd -> %zd.", key.c_str(), ipv.size(), alsv.size());
-                    ipv.resize(alsv.size());
+                    // No next hop specified, so just pop the label.
```

Comment:
Not sure what the intention is. I see two type of operations for a pop label:
1 - pop and forward on next hop
2 - pop and do another forwarding lookup

It's not the first here. Are you expecting the 2nd case has no nexthop. For this case in linux the nexthop would be loopback. I don't know how it's specified in FRR. But I'm not sure if this is the right assumption.

Reply:
How does this translate to ASIC programming since the nexthop is empty?

Reply:
Done